doc/cascadia/Keybindings-Arguments.md

```diff
@@ -0,0 +1,315 @@
+Currently, it's a `std::unordered_map<KeyChord, ShortcutAction, ...>`, which
+uses the `KeyChord` to lookup the `ShortcutAction`. We'll need to introduce a
+new type `ActionAndArgs`:
+
+```csharp
+runtimeclass ActionAndArgs
+{
+    ShortcutAction action;
+    IKeyBindingArgs args;
+}
+```
+
+and we'll change the map in `AppKeyBindings` to a `std::unordered_map<KeyChord,
+ActionAndArgs, ...>`.
+
+When we're parsing keybindings, we'll need to construct args for each of the
+events to go with each binding. When we find some key chord bound to a given
+Action, we'll construct the `IKeyBindingArgs` for that action. For many actions,
+these args will be an empty class. However, when we do find an action that needs
+additional parsing, `AppKeyBindingsSerialization` will do the extra work to
+parse the args for that action.
+
+Once the `IKeyBindingArgs` is built for the keybinding, we'll set it in
+`AppKeyBindings` with a updated `AppKeyBindings::SetKeyBinding` call.
+`SetKeyBinding`'s signature will be updated to take a `ActionAndArgs` instead.
+
+### Executing KeyBinding Actions with Arguments
+
+When we're handling a keybinding in `AppKeyBindings::_DoAction`, we'll trigger
+the event handlers with the `IKeyBindingArgs` we've stored in the map with the
+`ShortcutAction`.
+
+Then, in `App`, we'll handle each of these events. We set up lambdas as event
+handlers for each event in `App::_HookupKeyBindings`. In each of those
+functions, We'll inspect the `IKeyBindingArgs` parameter, and use args from its
+implementation to call callbacks in the `App` class. We will update `App` to
+have methods defined with the actual keybinding function signatures.
+
+Instead of:
+
+```c++
+    void App::_HookupKeyBindings(TerminalApp::AppKeyBindings bindings) noexcept
+    {
+        // ...
+        bindings.NewTabWithProfile([this](const auto index) { _OpenNewTab({ index }); });
+    }
+```
+
+The code will look like:
+
+```c++
+    void App::_HookupKeyBindings(TerminalApp::AppKeyBindings bindings) noexcept
+    {
+        // ...
+        bindings.NewTabWithProfile({ this, &App::_OpenNewTab });
+    }
+    // ...
+    void App::_OpenNewTab(const TerminalApp::AppKeyBindings& sender, const NewTabEventArgs& args)
+    {
+        auto profileIndex = args.ProfileIndex();
+        args.Handled(true);
+        // ...
+    }
+```
+
+### Handling Keybinding Events
+
+Commmon to all implementations of `IKeyBindingArgs` is the `Handled` property.
+This will let the app indicate if it was able to actually process a keybinding
+event or not. While in the large majority of cases, the events will all be
+marked handled, there are some scenarios where the Terminal will need to know if
+the event could not be performed. For example, in the case of the `copy` event,
+the Terminal is only capable of copying text if there's actually a selection
+active. If there isn't a selection active, the `App` should make sure to mark
+the event as not handled (`args.Handled(false)`).

```

Comment:
To be consistent with all other implementations of `Handled` we may just need to default to `false` and _make sure the app always sets `true`_

Reply:
Good point, I'll clear up the language here.